File: android/src/main/java/io/textile/rnmobile/ThreadsBridge.java
package io.textile.rnmobile;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import io.textile.pb.Model;
import io.textile.pb.QueryOuterClass;
import io.textile.pb.View;
import io.textile.textile.Textile;
import mobile.SearchHandle;

public class ThreadsBridge extends ReactContextBaseJavaModule {

    private static SearchHandle searchHandle;

    private Executor executor = Executors.newSingleThreadExecutor();

    public ThreadsBridge(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "ThreadsBridge";
    }

    @ReactMethod
    public void add(final String configStr, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    View.AddThreadConfig config = View.AddThreadConfig.parseFrom(Util.decode(configStr));
                    Model.Thread thread = Textile.instance().threads.add(config);
                    promise.resolve(Util.encode(thread.toByteArray()));
                }
                catch (Exception e) {
                    promise.reject("add", e);
                }
            }
        });
    }

    @ReactMethod
    public void addOrUpdate(final String threadStr, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Model.Thread thread = Model.Thread.parseFrom(Util.decode(threadStr));
                    Textile.instance().threads.addOrUpdate(thread);
                    promise.resolve(null);
                }
                catch (Exception e) {
                    promise.reject("addOrUpdate", e);
                }
            }
        });
    }

    @ReactMethod
    public void rename(final String threadId, final String name, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Textile.instance().threads.rename(threadId, name);
                    promise.resolve(null);
                }
                catch (Exception e) {
                    promise.reject("rename", e);
                }
            }
        });
    }

    @ReactMethod
    public void get(final String threadId, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Model.Thread thread = Textile.instance().threads.get(threadId);
                    promise.resolve(Util.encode(thread.toByteArray()));
                }
                catch (Exception e) {
                    promise.reject("get", e);
                }
            }
        });
    }

    @ReactMethod
    public void list(final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Model.ThreadList list = Textile.instance().threads.list();
                    promise.resolve(Util.encode(list.toByteArray()));
                }
                catch (Exception e) {
                    promise.reject("list", e);
                }
            }
        });
    }

    @ReactMethod
    public void peers(final String threadId, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Model.ContactList list = Textile.instance().threads.peers(threadId);
                    promise.resolve(Util.encode(list.toByteArray()));
                }
                catch (Exception e) {
                    promise.reject("peers", e);
                }
            }
        });
    }

    @ReactMethod
    public void remove(final String threadId, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    promise.resolve(Textile.instance().threads.remove(threadId));
                }
                catch (Exception e) {
                    promise.reject("remove", e);
                }
            }
        });
    }

    @ReactMethod
    public void snapshot(final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Textile.instance().threads.snapshot();
                    promise.resolve(null);
                }
                catch (Exception e) {
                    promise.reject("snapshot", e);
                }
            }
        });
    }

    @ReactMethod
    public void searchSnapshots(final String queryStr, final String optionsStr, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    if (ThreadsBridge.searchHandle != null) {
                        ThreadsBridge.searchHandle.cancel();
                    }
                    QueryOuterClass.ThreadSnapshotQuery query = QueryOuterClass.ThreadSnapshotQuery.parseFrom(Util.decode(queryStr));
                    QueryOuterClass.QueryOptions options = QueryOuterClass.QueryOptions.parseFrom(Util.decode(optionsStr));
                    ThreadsBridge.searchHandle = Textile.instance().threads.searchSnapshots(query, options);
                    promise.resolve(ThreadsBridge.searchHandle.getId());
                }
                catch (Exception e) {
                    promise.reject("searchSnapshots", e);
                }
            }
        });
    }

    @ReactMethod
    public void cancelSearchSnapshots(final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    if (ThreadsBridge.searchHandle != null) {
                        ThreadsBridge.searchHandle.cancel();
                        ThreadsBridge.searchHandle = null;
                    }
                    promise.resolve(null);
                }
                catch (Exception e) {
                    promise.reject("cancelSync", e);
                }
            }
        });
    }
}

File: ios/ThreadsBridge.m
#import "ThreadsBridge.h"

#if __has_include(<React/RCTBridge.h>)
#import <React/RCTBridge.h>
#elif __has_include(“RCTBridge.h”)
#import “RCTBridge.h”
#else
#import “React/RCTBridge.h”
#endif

#import <Textile/TextileApi.h>
#import "utils.h"

#define SYSTEM_VERSION_LESS_THAN(v) ([[[UIDevice currentDevice] systemVersion] compare:v options:NSNumericSearch] == NSOrderedAscending)

@interface ThreadsBridge ()

@property (atomic, strong) MobileSearchHandle *searchHandle;

@end

@implementation ThreadsBridge

RCT_EXPORT_MODULE();

- (dispatch_queue_t)methodQueue {
  return dispatch_queue_create("io.textile.TextileNodeQueue", DISPATCH_QUEUE_SERIAL);
}

RCT_EXPORT_METHOD(add:(NSString*)configStr resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  NSData *configData = [[NSData alloc] initWithBase64EncodedString:configStr options:0];
  AddThreadConfig *config = [[AddThreadConfig alloc] initWithData:configData error:&error];
  Thread *thread = [Textile.instance.threads add:config error:&error];
  fulfillWithResult([thread.data base64EncodedStringWithOptions:0], error, resolve, reject);
}

RCT_EXPORT_METHOD(addOrUpdate:(NSString*)threadStr resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  NSData *threadData = [[NSData alloc] initWithBase64EncodedString:threadStr options:0];
  Thread *thread = [[Thread alloc] initWithData:threadData error:&error];
  [Textile.instance.threads addOrUpdate:thread error:&error];
  fulfillWithResult(nil, error, resolve, reject);
}

RCT_EXPORT_METHOD(rename:(NSString*)threadId name:(NSString*)name resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  [Textile.instance.threads rename:threadId name:name error:&error];
  fulfillWithResult(nil, error, resolve, reject);
}

RCT_EXPORT_METHOD(get:(NSString*)threadId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  Thread *thread = [Textile.instance.threads get:threadId error:&error];
  fulfillWithResult([thread.data base64EncodedStringWithOptions:0], error, resolve, reject);
}

RCT_EXPORT_METHOD(list:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  ThreadList *list = [Textile.instance.threads list:&error];
  fulfillWithResult([list.data base64EncodedStringWithOptions:0], error, resolve, reject);
}

RCT_EXPORT_METHOD(peers:(NSString*)threadId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  ContactList *list = [Textile.instance.threads peers:threadId error:&error];
  fulfillWithResult([list.data base64EncodedStringWithOptions:0], error, resolve, reject);
}

RCT_EXPORT_METHOD(remove:(NSString*)id_ resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  NSString *result = [Textile.instance.threads remove:id_ error:&error];
  fulfillWithResult(result, error, resolve, reject);
}

RCT_EXPORT_METHOD(snapshot:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  [Textile.instance.threads snapshot:&error];
  fulfillWithResult(nil, error, resolve, reject);
}

RCT_EXPORT_METHOD(searchSnapshots:(NSString*)queryStr options:(NSString*)optionsStr resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  if (self.searchHandle) {
    [self.searchHandle cancel];
    self.searchHandle = nil;
  }
  NSError *error;
  NSData *queryData = [[NSData alloc] initWithBase64EncodedString:queryStr options:0];
  ThreadSnapshotQuery *query = [[ThreadSnapshotQuery alloc] initWithData:queryData error:&error];
  NSData *optionsData = [[NSData alloc] initWithBase64EncodedString:optionsStr options:0];
  QueryOptions *options = [[QueryOptions alloc] initWithData:optionsData error:&error];
  self.searchHandle = [Textile.instance.threads searchSnapshots:query options:options error:&error];
  fulfillWithResult(self.searchHandle.id_, error, resolve, reject);
}

RCT_EXPORT_METHOD(cancelSearchSnapshots:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  if (self.searchHandle) {
    [self.searchHandle cancel];
    self.searchHandle = nil;
  }
  resolve(nil);
}

@end

@implementation RCTBridge (ThreadsBridge)

- (ThreadsBridge *)threadsBridge {
  return [self moduleForClass:[ThreadsBridge class]];
}

@end
